File: editor/en/[42]Trapping Rain Water.java
// leetcode submit region begin(Prohibit modification and deletion)
class Solution {
  public int trap(int[] height) {
    int leftIndex = 0;
    int rightIndex = height.length - 1;
    int leftMaxHeight = 0;
    int rightMaxHeight = 0;
    int result = 0;

    while (leftIndex < rightIndex) {
      if (height[leftIndex] <= height[rightIndex]) {
        if (height[leftIndex] >= leftMaxHeight) {
          leftMaxHeight = height[leftIndex];
        } else {
          result += (leftMaxHeight - height[leftIndex]);
            }

        leftIndex++;
      } else {
        if (height[rightIndex] >= rightMaxHeight) {
          rightMaxHeight = height[rightIndex];
        } else {
          result += (rightMaxHeight - height[rightIndex]);
        }

        rightIndex--;
        }
    }

    return result;
  }
}
// leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[42]Trapping Rain Water.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def trap(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_left = 0
        max_right = 0
        trapped_water = 0

        while left < right:
            if height[left] <= height[right]:
                if height[left] >= max_left:
                    max_left = height[left]
                else:
                    trapped_water += max_left - height[left]
                left += 1
            else:
                if height[right] >= max_right:
                    max_right = height[right]
                else:
                    trapped_water += max_right - height[right]
                right -= 1

        return trapped_water

# leetcode submit region end(Prohibit modification and deletion)
